dump/match_infra.py

```diff
@@ -0,0 +1,287 @@
+            
+    def hget(self, db, key, field):
+        sp = self.sep(db)
+        tokens = key.split(sp)
+        key_ptrn = tokens[-1]
+        tokens.pop()
+        table = sp.join(tokens)
+        if table in self.db_driver and key_ptrn in self.db_driver[table] and field in self.db_driver[table][key_ptrn]:
+            return self.db_driver[table][key_ptrn][field]
+        return ""
+                        
+class MatchEngine:
+    
+    # Given a request obj, find its match in the redis
+    def fetch(self, req):
+        verbose_print(str(req))
+        template = self.__ret_template()
+        template['error']  = self.__validate_request(req)
+        if template['error']:
+            return self.__return_error(template)
+        
+        src = None
+        d_src = ""
+        if req.db:
+            d_src = req.db
+            src = RedisSource()
+        else:
+            d_src = req.file
+            src = JsonSource()
+        
+        if not src.connect(d_src, req.ns):
+            template['error']  = error_dict["CONN_ERR"]
+            return self.__return_error(template)
+        verbose_print("MatchRequest Checks Passed")
+        all_matched_keys = src.getKeys(req.db, req.table, req.key_pattern)
+        if not all_matched_keys or len(all_matched_keys) == 0:
+            template['error'] = error_dict["INV_PTTRN"]
+            return self.__return_error(template)
+        
+        filtered_keys = self.__filter_out_keys(src, req, all_matched_keys)
+        verbose_print("Filtered Keys:" + str(filtered_keys))
+        if not filtered_keys:
+            template['error'] = error_dict["NO_ENTRIES"]
+            return self.__return_error(template)
+        return self.__fill(src, req, filtered_keys)
+    
+    def __ret_template(self):
+        return {"error" : "", "keys" : [], "return_values" : {}}
+    
+    
+    def __return_error(self, template):
+        verbose_print("MatchEngine: \n" + template['error'])
+        return template
+    
+    def __validate_request(self, req):
+        
+        if not isinstance(req, MatchRequest):
+            return error_dict["INV_REQ"]
+            
+        if not(req.db) and not(req.file):
+            return error_dict["NO_SRC"]
+        
+        if req.db and req.file:
+            return error_dict["SRC_VAGUE"]
+        
+        if not req.db and os.path.exists(req.file):
+            try:
+                with open(req.file) as f:
+                    json.load(f)
+            except ValueError as e:
+                return error_dict["INV_JSON"] 
+        elif not req.db:
+            return error_dict["NO_FILE"]
+        
+        if not(req.file) and req.db not in SonicDBConfig.getDbList():
+            return error_dict["INV_DB"]
+        
+        if not req.table:
+            return error_dict["NO_TABLE"]
+        
+        if not req.key_pattern:
+            return error_dict["NO_KEY"]
+        
+        if not isinstance(req.return_fields, list):
```

Comment:
Why limit it to list, can't it be a generic iterable type?

Reply:
List or set might be valid here, if i would check this for a generic iterable type, a string or a dict might sneak in. Instead to simplify the design, would just be restricting this to list